### PR TITLE
Decrease a lock contention in PipelinedStageExecution

### PR DESCRIPTION
**Pull request by radek-starburst**
## Description
We observed that there is a high lock contention on `io.trino.execution.scheduler.PipelinedStageExecution`'s monitor.
Lock contention before:

new_master

Lock contention after:

Screenshot 2022-09-30 at 11 53 50

We measured the throughput (80 trino workers, 40 r5.4xlarge nodes, 64 concurrent queries) before/after

Throughput (queries/h) before:
**6000 queries / h**

Throughput (queries/h) after:
**6600 queries / h** (10% difference)

## Non-technical explanation
The Trino is able to process more queries within one hour.

## Release notes

( ) This is not user-visible and no release notes are required.
( ) Release notes are required, please propose a release note for me.
(* ) Release notes are required, with the following suggested text:
Increase throughput when running highly concurrent workloads on big Trino clusters

### CODE REVIEW

File: core/trino-main/src/main/java/io/trino/execution/scheduler/PipelinedStageExecution.java [hunk 1]
```diff
@@ -118,10 +118,8 @@
     // current stage task tracking
     @GuardedBy("this")
     private final Set<TaskId> allTasks = new HashSet<>();
-    @GuardedBy("this")
-    private final Set<TaskId> finishedTasks = new HashSet<>();
-    @GuardedBy("this")
-    private final Set<TaskId> flushingTasks = new HashSet<>();
+    private final Set<TaskId> finishedTasks = ConcurrentHashMap.newKeySet();
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/execution/scheduler/PipelinedStageExecution.java:121]: In first commit just add synchronized methods and get rid of synchronization in `updateTaskStatus`.
Then in second commit introduce opportunistic locking

File: core/trino-main/src/main/java/io/trino/execution/scheduler/PipelinedStageExecution.java [hunk 2]
```diff
@@ -367,11 +358,10 @@ private synchronized void updateTaskStatus(TaskStatus taskStatus)
                 fail(new TrinoException(GENERIC_INTERNAL_ERROR, "A task is in the ABORTED state but stage is " + stageState));
                 break;
             case FLUSHING:
-                flushingTasks.add(taskStatus.getTaskId());
+                newFlushingOrFinishedTaskObserved = addFlushingTask(taskStatus.getTaskId());
                 break;
             case FINISHED:
-                finishedTasks.add(taskStatus.getTaskId());
-                flushingTasks.remove(taskStatus.getTaskId());
+                newFlushingOrFinishedTaskObserved = addFinishedTask(taskStatus.getTaskId());
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/execution/scheduler/PipelinedStageExecution.java:364]: `newFlushingOrFinishedTaskObserved` is unused

File: core/trino-main/src/main/java/io/trino/execution/scheduler/PipelinedStageExecution.java [hunk 3]
```diff
@@ -380,15 +370,49 @@ private synchronized void updateTaskStatus(TaskStatus taskStatus)
...
+                return flushingTasks.add(taskId);
+            }
+        }
+        return false;
+    }
+
+    private boolean addFinishedTask(TaskId taskId)
+    {
+        boolean changed = false;
+        if (!finishedTasks.contains(taskId) || flushingTasks.contains(taskId)) {
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/execution/scheduler/PipelinedStageExecution.java:407]: just `if (!finishedTasks.contains(taskId))` is sufficient

File: core/trino-main/src/main/java/io/trino/execution/scheduler/PipelinedStageExecution.java [hunk 4]
```diff
@@ -380,15 +370,49 @@ private synchronized void updateTaskStatus(TaskStatus taskStatus)
...
+            synchronized (this) {
+                changed = finishedTasks.add(taskId);
+                changed |= flushingTasks.remove(taskId);
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/execution/scheduler/PipelinedStageExecution.java:410]: no need for `OR` here. Under synchronization section task cannot be finished, but still in `flushingTasks`

File: core/trino-main/src/main/java/io/trino/execution/scheduler/PipelinedStageExecution.java (hunk 1 above)
Comment (radek-starburst) [reply core/trino-main/src/main/java/io/trino/execution/scheduler/PipelinedStageExecution.java:121]: ok

File: core/trino-main/src/main/java/io/trino/execution/scheduler/PipelinedStageExecution.java (hunk 2 above)
Comment (radek-starburst) [reply core/trino-main/src/main/java/io/trino/execution/scheduler/PipelinedStageExecution.java:364]: It was used in the previous version. I wanted check why we observe to products tests failures:
`Test execution is not finished yet, a deadlock or hang is suspected` 
but we still observe that problem.

File: core/trino-main/src/main/java/io/trino/execution/scheduler/PipelinedStageExecution.java (hunk 3 above)
Comment (radek-starburst) [reply core/trino-main/src/main/java/io/trino/execution/scheduler/PipelinedStageExecution.java:407]: No, it is not (at least on my eye). There is a race condition here - I mean this method handles changing of task statuses. On my eye it is possible that we could observer a such order about specific `taskId`.
`FLUSHING`, `FINISHED`, `FLUSHING`, `FINISHED`. It could happen because of network latencies / request retrials. Then, we need a possibility to move the task from `FLUSHING` to `FINISHED` again. 
It is not something new - I mean there is exactly same issue in the fully synchronized version.

File: core/trino-main/src/main/java/io/trino/execution/scheduler/PipelinedStageExecution.java (hunk 4 above)
Comment (radek-starburst) [reply core/trino-main/src/main/java/io/trino/execution/scheduler/PipelinedStageExecution.java:410]: I would not say that (please look at my previous comment).

File: core/trino-main/src/main/java/io/trino/execution/scheduler/PipelinedStageExecution.java (hunk 3 above)
Comment (Dith3r) [reply core/trino-main/src/main/java/io/trino/execution/scheduler/PipelinedStageExecution.java:407]: But you don't test if the `FINISHED` task is set again to `FLUSHING`.

Comment (radek-starburst) [reply core/trino-main/src/main/java/io/trino/execution/scheduler/PipelinedStageExecution.java:407]: Yes, but the previous version (before the change) does not it as well.

File: core/trino-main/src/main/java/io/trino/execution/scheduler/PipelinedStageExecution.java [hunk 5]
```diff
@@ -380,13 +370,49 @@ private synchronized void updateTaskStatus(TaskStatus taskStatus)
...
             }
-            if (finishedTasks.containsAll(allTasks)) {
-                stateMachine.transitionToFinished();
+        }
+    }
+
+    private synchronized boolean isStageFlushing()
+    {
+        // to transition to flushing, there must be at least one flushing task, and all others must be flushing or finished.
+        return !flushingTasks.isEmpty()
```
Comment (Dith3r) [inline core/trino-main/src/main/java/io/trino/execution/scheduler/PipelinedStageExecution.java:387]: If there is a possibility of out-of-order execution then flushingTasks could end up as not empty. For each update of a task we will go through allTasks twice (one `isStageFlushing` and the second time in `isStageFinished`).

File: core/trino-main/src/main/java/io/trino/execution/scheduler/PipelinedStageExecution.java (hunk 3 above)
Comment (sopel39) [reply core/trino-main/src/main/java/io/trino/execution/scheduler/PipelinedStageExecution.java:409]: task shouldn't move back to `flushing` state. Therefore `finishedTasks.contains(taskId) && flushingTasks.contains(taskId)` is not a valid state

File: core/trino-main/src/main/java/io/trino/execution/scheduler/PipelinedStageExecution.java (hunk 5 above)
Comment (radek-starburst) [reply core/trino-main/src/main/java/io/trino/execution/scheduler/PipelinedStageExecution.java:387]: Yes, you right. Anyway, when I was rewriting the code I did my best to write more or less equivalent code. 
I mean, in the previous code there was same issue.

File: core/trino-main/src/main/java/io/trino/execution/scheduler/PipelinedStageExecution.java (hunk 3 above)
Comment (radek-starburst) [reply core/trino-main/src/main/java/io/trino/execution/scheduler/PipelinedStageExecution.java:409]: Yes, it shoud not. But, as we were discussing it is possbile that we process task statuses in the out of order manner.

File: core/trino-main/src/main/java/io/trino/execution/scheduler/PipelinedStageExecution.java [hunk 6]
```diff
@@ -380,13 +370,49 @@ private synchronized void updateTaskStatus(TaskStatus taskStatus)
...
+    private synchronized boolean isStageFinished()
+    {
+        return finishedTasks.containsAll(allTasks);
+    }
+
+    private boolean addFlushingTask(TaskId taskId)
+    {
+        if (!flushingTasks.contains(taskId)) {
+            synchronized (this) {
+                return flushingTasks.add(taskId);
```
[the remark this answers is not in the document]
Comment (radek-starburst) [reply core/trino-main/src/main/java/io/trino/execution/scheduler/PipelinedStageExecution.java:400]: Yes, to be safe for `out of order` case. I will do.

File: core/trino-main/src/main/java/io/trino/execution/scheduler/PipelinedStageExecution.java [hunk 7]
```diff
@@ -380,15 +370,49 @@ private synchronized void updateTaskStatus(TaskStatus taskStatus)
...
+    private synchronized boolean isStageFlushing()
+    {
+        // to transition to flushing, there must be at least one flushing task, and all others must be flushing or finished.
+        return !flushingTasks.isEmpty()
+                && allTasks.stream().allMatch(taskId -> finishedTasks.contains(taskId) || flushingTasks.contains(taskId));
+    }
+
+    private synchronized boolean isStageFinished()
+    {
+        return finishedTasks.containsAll(allTasks);
```
[the remark this answers is not in the document]
Comment (radek-starburst) [reply core/trino-main/src/main/java/io/trino/execution/scheduler/PipelinedStageExecution.java:393]: Yes, you right. `containsAll` is just more readable and it was used before.

[the remark this answers is not in the document]
Comment (Dith3r) [reply core/trino-main/src/main/java/io/trino/execution/scheduler/PipelinedStageExecution.java:393]: containsAll is O(N) and we know that this cannot succeed if the source set size is bigger than the target set. With the current implementation, we are looping through allTask twice. Avoiding this could reduce the time needed in the synchronized section.

File: core/trino-main/src/main/java/io/trino/execution/scheduler/PipelinedStageExecution.java [hunk 8]
```diff
@@ -380,13 +370,50 @@ private synchronized void updateTaskStatus(TaskStatus taskStatus)
...
+    }
+
+    private boolean addFlushingTask(TaskId taskId)
+    {
+        if (!flushingTasks.contains(taskId) && !finishedTasks.contains(taskId)) {
+            synchronized (this) {
```
Comment (lukasz-stec) [inline core/trino-main/src/main/java/io/trino/execution/scheduler/PipelinedStageExecution.java:399]: do you need synchronized (this)?

File: core/trino-main/src/main/java/io/trino/execution/scheduler/PipelinedStageExecution.java [hunk 9]
```diff
@@ -380,13 +370,50 @@ private synchronized void updateTaskStatus(TaskStatus taskStatus)
...
+                if (!finishedTasks.contains(taskId)) {
```
Comment (lukasz-stec) [inline core/trino-main/src/main/java/io/trino/execution/scheduler/PipelinedStageExecution.java:400]: is this check needed? wouldn't using only return `flushingTasks.add(taskId); `work?

Comment (radek-starburst) [reply core/trino-main/src/main/java/io/trino/execution/scheduler/PipelinedStageExecution.java:400]: Yes - we do not want to add task to the `flushingTasks` if the task is finished.

File: core/trino-main/src/main/java/io/trino/execution/scheduler/PipelinedStageExecution.java (hunk 8 above)
Comment (radek-starburst) [reply core/trino-main/src/main/java/io/trino/execution/scheduler/PipelinedStageExecution.java:399]: Yes, I need. I do not want to change `flushingTask` when `isStageFlushing` is reading from this set.

File: core/trino-main/src/main/java/io/trino/execution/scheduler/PipelinedStageExecution.java [hunk 10]
```diff
@@ -225,10 +223,10 @@ public synchronized void schedulingComplete()
             return;
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/execution/scheduler/PipelinedStageExecution.java:223]: `synchronized` is redundant now in `schedulingComplete`

File: core/trino-main/src/main/java/io/trino/execution/scheduler/PipelinedStageExecution.java [hunk 11]
```diff
@@ -237,13 +235,6 @@ public synchronized void schedulingComplete()
         }
     }
 
-    private synchronized boolean isFlushing()
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/execution/scheduler/PipelinedStageExecution.java:240]: nit: that should be separate commit (moving of method)

File: core/trino-main/src/main/java/io/trino/execution/scheduler/PipelinedStageExecution.java [hunk 12]
```diff
@@ -380,13 +370,50 @@ private synchronized void updateTaskStatus(TaskStatus taskStatus)
             if (taskState == TaskState.RUNNING) {
                 stateMachine.transitionToRunning();
             }
-            if (isFlushing()) {
-                stateMachine.transitionToFlushing();
+            if (newFlushingOrFinishedTaskObserved) {
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/execution/scheduler/PipelinedStageExecution.java:373]: add a comment, e.g: `avoid extra synchronization if no flushing or finished task was observed`

File: core/trino-main/src/main/java/io/trino/execution/scheduler/PipelinedStageExecution.java [hunk 13]
```diff
@@ -380,13 +370,50 @@ private synchronized void updateTaskStatus(TaskStatus taskStatus)
...
+                if (isStageFinished()) {
+                    stateMachine.transitionToFinished();
+                }
             }
-            if (finishedTasks.containsAll(allTasks)) {
-                stateMachine.transitionToFinished();
+        }
+    }
+
+    private synchronized boolean isStageFlushing()
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/execution/scheduler/PipelinedStageExecution.java:384]: did this method change?

File: core/trino-main/src/main/java/io/trino/execution/scheduler/PipelinedStageExecution.java (hunk 9 above)
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/execution/scheduler/PipelinedStageExecution.java:400]: add a comment why we need this `!finishedTasks.contains(taskId))` check (out-of-order events)

File: core/trino-main/src/main/java/io/trino/execution/scheduler/PipelinedStageExecution.java [hunk 14]
```diff
@@ -380,13 +370,50 @@ private synchronized void updateTaskStatus(TaskStatus taskStatus)
...
+            }
+        }
+        return false;
+    }
+
+    private boolean addFinishedTask(TaskId taskId)
+    {
+        if (!finishedTasks.contains(taskId)) {
+            synchronized (this) {
+                flushingTasks.remove(taskId);
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/execution/scheduler/PipelinedStageExecution.java:412]: nit: if you change the order
```
boolean added = finishedTasks.add(taskId);
flushingTasks.remove(taskId);
return added;
```
then there won't be a race condition with opportunistic lock in `addFlushingTask`:
```
if (!flushingTasks.contains(taskId) && !finishedTasks.contains(taskId))) {
  synchronized(this) {
   ... 
  }
}
```

**Review by sopel39 — COMMENTED** [review]
lgtm % comments

**Review by sopel39 — APPROVED** [review]

File: core/trino-main/src/main/java/io/trino/execution/scheduler/PipelinedStageExecution.java (hunk 13 above)
Comment (radek-starburst) [reply core/trino-main/src/main/java/io/trino/execution/scheduler/PipelinedStageExecution.java:384]: No, only the name was changed.

**Review by Dith3r — APPROVED** [review]
LGTM.

**Review by sopel39 — APPROVED** [review]

**Comment by sopel39** [comment]
There are some test failures. Could you rerun?

**Comment by radek-starburst** [comment]
> There are some test failures. Could you rerun?

Yes, these failures are not relevant. I am going to rerun but before I want to see result of product tests. After that, I rerun.

File: core/trino-main/src/main/java/io/trino/execution/scheduler/PipelinedStageExecution.java [hunk 15]
```diff
@@ -380,13 +370,54 @@ private synchronized void updateTaskStatus(TaskStatus taskStatus)
             if (taskState == TaskState.RUNNING) {
                 stateMachine.transitionToRunning();
             }
-            if (isFlushing()) {
-                stateMachine.transitionToFlushing();
+            // avoid extra synchronization if no new flushing or finished task was observed
+            if (newFlushingOrFinishedTaskObserved) {
```
Comment (arhimondr) [inline core/trino-main/src/main/java/io/trino/execution/scheduler/PipelinedStageExecution.java:374]: The `stageState` is read before a task is added to the `finishedTasks` set. I think this breaks happens-before assumptions. It is possible that `State stageState = stateMachine.getState();` is run when the state is not yet `SCHEDULED` so the `stateMachine.transitionToFinished();` is not run while at the same time the `schedulingComplete()` transitions the stage to `SCHEDULED` but doesn't see all tasks being finished.

Comment (sopel39) [reply core/trino-main/src/main/java/io/trino/execution/scheduler/PipelinedStageExecution.java:374]: Yup. You are correct.
